### PR TITLE
Add validations for K8s resource fields

### LINKED ISSUE #1405: Validate arguments before applying them
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Validate obvious arguments before calling SDK. It is probably better to validate on both sides (SDK and Terraform provider).

This is an example of a failure which took 2 minutes:

```
* azurerm_kubernetes_cluster.test: Long running operation terminated with status 'Failed': Code="InvalidParameter" Message="The Admin Username specified is not allowed."
```

### New or Affected Resource(s)

* azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
resource "azurerm_kubernetes_cluster" "test" {
  linux_profile {
    admin_username = "user"
    # ...
  }
}

```

### PR DESCRIPTION
**Pull request by metacpp**
This PR adds some validation methods for K8s resource fields.

Fixes #1405.

### CODE REVIEW

File: azurerm/helpers/validate/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		"AdminUserName must start with alphabet and/or continue with alphanumeric characters, underscores, hyphens.")
+}
+
+func KubernetesCidr() schema.SchemaValidateFunc {
+	return validation.StringMatch(
+		regexp.MustCompile(`^([0-9]{1,3}\.){3}[0-9]{1,3}(\/([0-9]|[1-2][0-9]|3[0-2]))?$`),
+		"CIDR must start with IPV4 address and/or slash, number of bits (0-32) as prefix. Example: 127.0.0.1/8.")
+}
+
+func KubernetesDNSServiceIP() schema.SchemaValidateFunc {
```
Comment (katbyte) [inline azurerm/helpers/validate/kubernetes.go]: Is there a reason you did not just use `validate.IPv4Address`?

File: azurerm/helpers/validate/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func KubernetesAdminUserName() schema.SchemaValidateFunc {
+	return validation.StringMatch(
+		regexp.MustCompile(`^[A-Za-z][-A-Za-z0-9_]*$`),
+		"AdminUserName must start with alphabet and/or continue with alphanumeric characters, underscores, hyphens.")
+}
+
+func KubernetesCidr() schema.SchemaValidateFunc {
```
Comment (katbyte) [inline azurerm/helpers/validate/kubernetes.go]: This should be added to `helpers/validate/network.go` as it is not kubernetes specific

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 3]
```diff
@@ -135,9 +138,10 @@ func resourceArmKubernetesCluster() *schema.Resource {
 						},
 
 						"vnet_subnet_id": {
-							Type:     schema.TypeString,
-							Optional: true,
-							ForceNew: true,
+							Type:         schema.TypeString,
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we use `azure.ValidateResourceIDOrEmpty` here"

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 4]
```diff
@@ -225,8 +231,9 @@ func resourceArmKubernetesCluster() *schema.Resource {
 										Required: true,
 									},
 									"log_analytics_workspace_id": {
-										Type:     schema.TypeString,
-										Required: true,
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we use `azure.ValidateResourceIDOrEmpty` here"

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 5]
```diff
@@ -302,31 +312,35 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
 						"dns_service_ip": {
-							Type:     schema.TypeString,
-							Optional: true,
-							Computed: true,
-							ForceNew: true,
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ForceNew:     true,
+							ValidateFunc: validate.KubernetesDNSServiceIP(),
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we use `validate.IPv4Address` instead?

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 6]
```diff
@@ -353,30 +367,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"client_app_id": {
-										Type:     schema.TypeString,
-										Required: true,
-										ForceNew: true,
+										Type:         schema.TypeString,
+										Required:     true,
+										ForceNew:     true,
+										ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we use `azure.ValidateResourceID` here"

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 7]
```diff
@@ -353,30 +367,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
 									},
 
 									"server_app_id": {
-										Type:     schema.TypeString,
-										Required: true,
-										ForceNew: true,
+										Type:         schema.TypeString,
+										Required:     true,
+										ForceNew:     true,
+										ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we use `azure.ValidateResourceID` here?

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 8]
```diff
@@ -353,30 +367,34 @@ func resourceArmKubernetesCluster() *schema.Resource {
...
 										// this can be sourced from the client config if it's not specified
-										Type:     schema.TypeString,
-										Optional: true,
-										Computed: true,
-										ForceNew: true,
+										Type:         schema.TypeString,
+										Optional:     true,
+										Computed:     true,
+										ForceNew:     true,
+										ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster.go]: Could we validate this with `validate.UUIDorEmpty`

File: azurerm/resource_arm_kubernetes_cluster_test.go [hunk 9]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/terraform"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
 )
 
 func TestAzureRMKubernetesCluster_agentPoolName(t *testing.T) {
```
Comment (katbyte) [inline azurerm/resource_arm_kubernetes_cluster_test.go]: Could we move this to the test file next to the function's implementation? Doesn't make much sense to be here

Comment (metacpp) [reply azurerm/resource_arm_kubernetes_cluster_test.go]: Thanks for pointing it out. I've merged the tests cases into `TestKubernetesAgentPoolName` in `kubernetes_test.go`.

**Comment by metacpp** [comment]
@katbyte I updated the PR through merging the tests into one.

**Comment by katbyte** [comment]
Thanks @metacpp, LGTM now 👍

**Pull request merged**